bootstrap/internal/controller/infraenv_controller.go

```diff
@@ -68,45 +95,66 @@ func (r *InfraEnvReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
-	log.WithValues("infra_env", infraEnv.Name, "infra_env_namespace", infraEnv.Namespace)
+	log = log.WithValues("infra_env", infraEnv.Name, "infra_env_namespace", infraEnv.Namespace)
 
 	if infraEnv.Status.ISODownloadURL == "" {
 		log.V(logutil.TraceLevel).Info("image URL not available yet")
-		return ctrl.Result{}, nil
+		// requeue so we'll make sure we'll have the ISO available.
+		// NOTE: We should not need this, as InfraEnv should notify us when changing status
+		return ctrl.Result{Requeue: true, RequeueAfter: retryAfter}, nil
+	}
+	err := r.attachISOToAgentBootstrapConfigs(ctx, infraEnv)
+	if err != nil {
+		return ctrl.Result{Requeue: true, RequeueAfter: time.Second * 5}, err
 	}
-	return ctrl.Result{}, r.attachISOToAgentBootstrapConfigs(ctx, infraEnv)
+	return ctrl.Result{}, nil
+
 }
 
 func (r *InfraEnvReconciler) attachISOToAgentBootstrapConfigs(ctx context.Context, infraEnv *aiv1beta1.InfraEnv) error {
-	clusterName, ok := infraEnv.Labels[clusterv1.ClusterNameLabel]
+	log := ctrl.LoggerFrom(ctx)
+
+	/*clusterName, ok := infraEnv.Labels[clusterv1.ClusterNameLabel]
 	if !ok {
 		return nil
 	}
-
+	*/
 	agentBootstrapConfigs := &bootstrapv1alpha1.AgentBootstrapConfigList{}
-	if err := r.Client.List(ctx, agentBootstrapConfigs, client.MatchingLabels{clusterv1.ClusterNameLabel: clusterName}); err != nil {
+	/*if err := r.Client.List(ctx, agentBootstrapConfigs, client.MatchingLabels{clusterv1.ClusterNameLabel: clusterName}); err != nil {
+		return errors.Wrap(err, "failed to list agent bootstrap configs")
+	}*/
+	if err := r.Client.List(ctx, agentBootstrapConfigs, client.MatchingFields{abcInfraEnvRefFieldName: infraEnv.Name, abcInfraEnvRefFieldNamespace: infraEnv.Namespace}); err != nil {
 		return errors.Wrap(err, "failed to list agent bootstrap configs")
 	}
 
+	log.V(logutil.TraceLevel).Info("listing agentBoostrapConfigs", "items found", len(agentBootstrapConfigs.Items), "LIST AgentBootstrapConfigs", agentBootstrapConfigs)
 	downloadURL, err := r.getISOURL(ctx, infraEnv.Status.ISODownloadURL)
+	log.V(logutil.TraceLevel).Info("ISO URL from INFRAENV", "ISO URL", downloadURL)
+
 	if err != nil {
+		log.V(logutil.TraceLevel).Info("error retrieving downloadURL from infraenv", "infraenv", infraEnv)
 		return err
 	}
 
+	var errorIfSkipped error = nil
 	for _, agentBootstrapConfig := range agentBootstrapConfigs.Items {
-		if agentBootstrapConfig.Status.InfraEnvRef == nil ||
-			(agentBootstrapConfig.Status.InfraEnvRef != nil &&
-				agentBootstrapConfig.Status.InfraEnvRef.Name != infraEnv.Name) {
+		if agentBootstrapConfig.Status.InfraEnvRef == nil {
+			log.V(logutil.TraceLevel).Info("skipping agentbootstrapconfig because no infraenv ref", "abc", agentBootstrapConfig.Name)
+			errorIfSkipped = errors.New("skipped an infraenv, need to retry") // infraenv would change and retrigger though
+			continue
+		}
+		if agentBootstrapConfig.Status.InfraEnvRef.Name != infraEnv.Name {
```

Comment:
Doesn't this https://github.com/openshift-assisted/cluster-api-agent/pull/55/files#diff-ac76a30087a84306870597bf5ad69a31f19e4bb602100a6dd9fda7f53418e443R126 already ensure we're only getting AgentBootstrapConfigs that have `Status.InfraEnvRef` set to this infra env?

Reply:
yes, good catch

Reply:
https://github.com/openshift-assisted/cluster-api-agent/pull/56